core/output/smartcard.js

```diff
@@ -0,0 +1,318 @@
+      const { cardHandle, activeProtocol } = await this._connect(this.context, this.readers[0]);
+      this.cardHandle = cardHandle;
+      this.activeProtocol = activeProtocol;
+    }
+  }
+
+  async powerOff() {
+    if (this.context && this.cardHandle) {
+      await this._disconnect(this.context, this.cardHandle);
+      await this._releaseContext(this.context);
+    }
+
+    this.context = null;
+    this.cardHandle = null;
+    this.cardAtr = null;
+    this.activeProtocol = null;
+  }
+
+  async transmit(apdu) {
+    try {
+      await this._beginTransaction();
+    } catch (error) {}
+
+    try {
+      this.lastTransmitAt = Date.now();
+      return await this._transmit(apdu);
+    } catch (error) {
+      this.lastTransmitAt = null;
+      throw error;
+    } finally {
+      await this._endTransaction();
+    }
+  }
+
+  async _establishContext() {
+    return await this._callExtension("establish_context", 0).then(([status, context]) => context);
+  }
+
+  async _releaseContext(context) {
+    return await this._callExtension("release_context", context).then(([status]) => status);
+  }
+
+
+  async _listReaders(context) {
+    return await this._callExtension("list_readers", context).then(([status, readers]) => {
+      return Array.isArray(readers) ? readers : readers.split(",").filter(Boolean);
+    });
+  }
+
+  async _getStatusChange(context, reader) {
+    return await this._callExtension("get_status_change", context, 0, 1, 0, 0, reader).then(
+      ([status, readerCount, currentState, eventState, atr]) => ({
+        status,
+        readerCount,
+        currentState,
+        eventState,
+        atr,
+      })
+    );
+  }
+
+  async _connect(context, reader) {
+    return await this._callExtension("connect", context, 2, 3, reader).then(
+      ([status, cardContext, cardHandle, activeProtocol]) => ({
+        cardHandle,
+        activeProtocol,
+      })
+    );
+  }
+
+  async _disconnect(context, cardHandle) {
+    return await this._callExtension("disconnect", context, cardHandle, 0).then(([status]) => status);
+  }
+
+  async _beginTransaction() {
+    return await this._callExtension("begin_transaction", this.context, this.cardHandle).then(([status]) => status);
+  }
+
+  async _transmit(apdu) {
+    return await this._callExtension("transmit", this.context, this.cardHandle, this.activeProtocol, toHex(apdu)).then(
+      ([status, context, card, protocol, response]) => fromHex(response)
+    );
+  }
+
+  async _endTransaction(disposition = 0) {
+    return await this._callExtension("end_transaction", this.context, this.cardHandle, disposition).then(
+      ([status]) => status
+    );
+  }
+
+  async _callExtension(name, ...args) {
+    return new Promise((resolve, reject) => {
+      const deviceId = "smartcard-relay";
+      const completionId = Date.now().toString() + Math.random().toString(36);
+
+      const message = {
+        deviceId,
+        completionId,
+        type: name,
+        args: args.join(","),
```

Comment:
I don't much about these communications, so this may not be a concern at all, but wanted to ask... Are we sure that all args will never have a comma in them? If a cardHandle or context could have a comma, it may result in undefined behavior.

Reply:
@mmcclaskey AFAIK, there should be no commas. We also verify the arguments on per-message basis in the chrome extension so, in the worst case scenario there wouldn't be any undefined behaviour, only errors.